homeassistant/config.py

```diff
@@ -212,6 +215,9 @@ def _filter_bad_internal_external_urls(conf: dict) -> dict:
             vol.Optional(CONF_ALLOWLIST_EXTERNAL_DIRS): vol.All(
                 cv.ensure_list, [vol.IsDir()]  # pylint: disable=no-value-for-parameter
             ),
+            vol.Optional(CONF_PRELOAD_TRANSLATIONS): vol.All(
+                cv.ensure_list, [cv.string]
+            ),
```

Comment:
I talked this a bit over with some other core maintainers.

We are open to a translation function in our jinja templates, but we should languages on first use and don't want to add a core configuration option for this.

So for now, we should remove this part from this PR. For the future we could do things like: Check the languages used by the users on this system and preload those, or even better, we might need a system-wide-default language in general.

Anyways, in either case, we like to keep that all out of scope for this PR.

Reply:
Thanks for your review!

Without preloading part translations are loaded during first access to frontend after start of HA. Should I add loading to first use of `state_translated` as well? I have tried to achieve it, but I never got it to work because translations loading is async.

Reply:
> Should I add loading to first use of state_translated as well?

We'll have to right? In theory, you could be using a language that isn't pre-loaded at all.

> I have tried to achieve it, but I never got it to work because translation loading is async.

I guess you'd could use `asyncio.run_coroutine_threadsafe()` to handle the call to `async_get_translations` in a sync context.

Reply:
> We'll have to right? In theory, you could be using a language that isn't pre-loaded at all.

That's true. In this case a state of the entity would be returned

> I guess you'd could use `asyncio.run_coroutine_threadsafe()` to handle the call to `async_get_translations` in a sync context.

As far as I remember I have tried it and it resulted in a deadlock... I can try it again though to make myself sure

Reply:
I have tried it again and it behaves in a quite weird way. My tests with dev tools template evaluator concluded that `StateTranslated.__call__` method is called 3 times, first time from some numbered thread (this time output is calculated correctly, but not returned to frontend) and then 2 times from `MainThread` (both of them result in timeout). Does anybody have any idea how to fix it?

Code:
```python
translations = asyncio.run_coroutine_threadsafe(async_get_translations(self._hass, language, "state", [domain]), self._hass.loop).result(5)
```

Reply:
@frenck do you have any idea/suggestion how to fix it?

Reply:
The problem still persists, even with much simplified version. Output is printed correctly, but template evaluation results in timeout nevertheless.

```python
class StateTranslated:

    def __call__(self, entity_id: str, language: str) -> str | None:
        output = asyncio.run_coroutine_threadsafe(self.get_value(entity_id, language), self._hass.loop).result(5)
        _LOGGER.debug(f"output: {output}")
        return output

    async def get_value(self, entity_id: str, language: str) -> str | None:
        await asyncio.sleep(1)
        return f"{entity_id} in {language}"
```

Reply:
Looked at it now, I don't see a way to make it work at this point.

Reply:
@frenck Do you think it would be possible to go along with approach similar to original one (preload translations at HA start)? Maybe using some other way to achieve it instead of yaml config

Reply:
I thought about adding a way to load translations without `async`, but it would require a lot of changes and it is preferred to execute IO in async way.

I'm out of ideas at this moment.

Reply:
Home Assistant has a language in its core config. That could allow for some smart preloading maybe.

Reply:
@balloob in this case it might be possible to drop `language` parameter of `state_translated` and just support the language selected in the settings. Would it be an acceptable solution?

Reply:
Yes. I think that would clean up like 90% of this PR too :)